File: src/StockBalancer.Core/Balancer.cs
namespace StockBalancer.Core;

public class Balancer : IBalancer
{
    public Portfolio Calculate(Portfolio portfolio)
    {
        var totalValue = portfolio.Positions.Sum(p => p.Quantity * p.Price);
        var totalAvailable = totalValue + portfolio.Balance;
        var newBalance = portfolio.Balance;

        var targetValues = portfolio.Positions.ToDictionary(
            p => p.Ticker,
            p => totalAvailable * (p.TargetPercentage / 100m)
        );

        var currentValues = portfolio.Positions.ToDictionary(
            p => p.Ticker,
            p => p.Quantity * p.Price
        );

        var newPositions = portfolio.Positions.Select(p => new Position
        {
            Ticker = p.Ticker, Quantity = p.Quantity, Price = p.Price, TargetPercentage = p.TargetPercentage
        }).ToList();

        bool purchased;
        do
        {
            purchased = false;
            var buyOptions = newPositions
                .Select(p => new { p.Ticker, p.Price, Deficit = targetValues[p.Ticker] - currentValues[p.Ticker] })
                .Where(x => x.Deficit > 0 && newBalance >= x.Price)
                .OrderByDescending(x => x.Deficit)
                .ToList();

            foreach (var option in buyOptions)
            {
                if (newBalance < option.Price)
                {
                    continue;
                }

                var position = newPositions.First(p => p.Ticker == option.Ticker);
                position.Quantity++;
                newBalance -= option.Price;
                currentValues[option.Ticker] += option.Price;
                purchased = true;
            }
        } while (purchased);

        return new Portfolio(newBalance, newPositions.ToArray());
    }
}

File: src/ui/services/balancer/Balancer.ts
import { IBalancer } from "./IBalancer.ts";
import { Portfolio } from "../../types/Portfolio.ts";
import { Position } from "../../types/Position.ts";

export class Balancer implements IBalancer {
  calculate(portfolio: Portfolio): Portfolio {
    const totalValue = portfolio.positions.reduce(
      (sum, p) => sum + p.quantity * p.price,
      0,
    );
    const totalAvailable = totalValue + portfolio.balance;
    let newBalance = portfolio.balance;

    const targetValues: Record<string, number> = {};
    portfolio.positions.forEach((p) => {
      targetValues[p.ticker] = totalAvailable * (p.target / 100);
    });

    const currentValues: Record<string, number> = {};
    portfolio.positions.forEach((p) => {
      currentValues[p.ticker] = p.quantity * p.price;
    });

    const newPositions: Position[] = portfolio.positions.map((p) => ({
      ticker: p.ticker,
      quantity: p.quantity,
      price: p.price,
      target: p.target,
    }));

    let purchased: boolean;
    do {
      purchased = false;

      const buyOptions = newPositions
        .map((p) => ({
          ticker: p.ticker,
          price: p.price,
          deficit: targetValues[p.ticker] - currentValues[p.ticker],
        }))
        .filter((x) => x.deficit > 0 && newBalance >= x.price)
        .sort((a, b) => b.deficit - a.deficit);

      for (const option of buyOptions) {
        if (newBalance < option.price) {
          continue;
        }

        const position = newPositions.find((p) => p.ticker === option.ticker);
        if (position) {
          position.quantity++;
          newBalance -= option.price;
          currentValues[option.ticker] += option.price;
          purchased = true;
        }
      }
    } while (purchased);

    return { balance: newBalance, positions: newPositions };
  }
}
